src/analysis/normed_space/lp_space.lean

```diff
@@ -0,0 +1,516 @@
+      { simpa using eventually_lt_of_tendsto_lt (by norm_num : (0:ℝ) < 1)
+          hf'.tendsto_cofinite_zero },
+      exact H.subset (λ i hi, real.one_le_rpow hi hq.le) },
+    { show ∀ i, ¬ (|∥f i∥ ^ p.to_real| ≤ ∥f i∥ ^ q.to_real) → 1 ≤ ∥f i∥,
+      intros i hi,
+      have : 0 ≤ ∥f i∥ ^ p.to_real := real.rpow_nonneg_of_nonneg (norm_nonneg _) p.to_real,
+      simp only [abs_of_nonneg, this] at hi,
+      contrapose! hi,
+      exact real.rpow_le_rpow_of_exponent_ge' (norm_nonneg _) hi.le hq.le hpq' } }
+end
+
+lemma add {f g : Π i, E i} (hf : mem_ℓp f p) (hg : mem_ℓp g p) : mem_ℓp (f + g) p :=
+begin
+  rcases p.trichotomy with rfl | rfl | hp,
+  { apply mem_ℓp_zero,
+    refine (hf.finite_dsupport.union hg.finite_dsupport).subset _,
+    intros i,
+    simp only [pi.add_apply, ne.def, set.mem_union_eq, set.mem_set_of_eq],
+    contrapose!,
+    rintros ⟨hf', hg'⟩,
+    simp [hf', hg'] },
+  { apply mem_ℓp_infty,
+    obtain ⟨A, hA⟩ := hf.bdd_above,
+    obtain ⟨B, hB⟩ := hg.bdd_above,
+    refine ⟨A + B, _⟩,
+    rintros a ⟨i, rfl⟩,
+    exact le_trans (norm_add_le _ _) (add_le_add (hA ⟨i, rfl⟩) (hB ⟨i, rfl⟩)) },
+  apply mem_ℓp_gen hp,
+  let C : ℝ := if p.to_real < 1 then 1 else 2 ^ (p.to_real - 1),
+  refine summable_of_nonneg_of_le _ _ (((hf.summable hp).add (hg.summable hp)).mul_left C),
+  { exact λ b, real.rpow_nonneg_of_nonneg (norm_nonneg (f b + g b)) p.to_real },
+  { intros i,
+    refine (real.rpow_le_rpow (norm_nonneg _) (norm_add_le _ _) hp.le).trans _,
+    dsimp [C],
+    split_ifs,
+    { simpa using nnreal.coe_le_coe.2 (nnreal.rpow_add_le_add_rpow (∥f i∥₊) (∥g i∥₊) hp h.le) },
+    { let F : fin 2 → ℝ≥0 := ![∥f i∥₊, ∥g i∥₊],
+      have : ∀ i, (0:ℝ) ≤ F i := λ i, (F i).coe_nonneg,
+      simp only [not_lt] at h,
+      simpa [F, fin.sum_univ_succ] using
+        real.rpow_sum_le_const_mul_sum_rpow_of_nonneg (finset.univ : finset (fin 2)) h
+        (λ i _, (F i).coe_nonneg) } }
+end
+
+lemma sub {f g : Π i, E i} (hf : mem_ℓp f p) (hg : mem_ℓp g p) : mem_ℓp (f - g) p :=
+by { rw sub_eq_add_neg, exact hf.add hg.neg }
+
+lemma finset_sum {ι} (s : finset ι) {f : ι → Π i, E i} (hf : ∀ i ∈ s, mem_ℓp (f i) p) :
+  mem_ℓp (λ a, ∑ i in s, f i a) p :=
+begin
+  haveI : decidable_eq ι := classical.dec_eq _,
+  revert hf,
+  refine finset.induction_on s _ _,
+  { simp only [zero_mem_ℓp', finset.sum_empty, implies_true_iff], },
+  { intros i s his ih hf,
+    simp only [his, finset.sum_insert, not_false_iff],
+    exact (hf i (s.mem_insert_self i)).add (ih (λ j hj, hf j (finset.mem_insert_of_mem hj))), },
+end
+
+section normed_space
+
+variables {𝕜 : Type*} [normed_field 𝕜] [Π i, normed_space 𝕜 (E i)]
+
+lemma const_smul {f : Π i, E i} (hf : mem_ℓp f p) (c : 𝕜) : mem_ℓp (c • f) p :=
+begin
+  rcases p.trichotomy with rfl | rfl | hp,
+  { apply mem_ℓp_zero,
+    refine hf.finite_dsupport.subset _,
+    intros i,
+    simp only [ne.def, set.mem_set_of_eq, pi.smul_apply],
+    contrapose!,
+    intros hf',
+    simp [hf'] },
```

Comment:
I may have taken this too far but this can be golfed to:
```suggestion
    refine hf.finite_dsupport.subset (λ i, (_ : ¬c • f i = 0 → ¬f i = 0)),
    exact not_imp_not.mpr (λ hf', hf'.symm ▸ (smul_zero c)) },
```